// @sap-cloud-sdk/generator/dist/complex-type/interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.complexTypeInterface = void 0;
const ts_morph_1 = require("ts-morph");
const typedoc_1 = require("../typedoc");
/**
 * @internal
 */
function complexTypeInterface(complexType) {
    return {
        kind: ts_morph_1.StructureKind.Interface,
        name: `${complexType.typeName}<DeSerializersT extends DeSerializers = DefaultDeSerializers>`,
        isExported: true,
        properties: properties(complexType),
        docs: [(0, typedoc_1.addLeadingNewline)(complexType.typeName)]
    };
}
exports.complexTypeInterface = complexTypeInterface;
function properties(complexType) {
    return complexType.properties.map(prop => property(prop));
}
function property(prop) {
    return {
        kind: ts_morph_1.StructureKind.PropertySignature,
        name: prop.instancePropertyName,
        type: `DeserializedType<DeSerializersT, '${prop.edmType}'>`,
        hasQuestionToken: prop.nullable,
        docs: [
            (0, typedoc_1.addLeadingNewline)((0, typedoc_1.getPropertyDescription)(prop, {
                nullable: prop.nullable,
                maxLength: prop.maxLength
            }))
        ]
    };
}
