// @sap-cloud-sdk/generator/dist/complex-type/namespace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.complexTypeNamespace = void 0;
const util_1 = require("@sap-cloud-sdk/util");
const ts_morph_1 = require("ts-morph");
/**
 * @internal
 */
function complexTypeNamespace(complexType) {
    return {
        kind: ts_morph_1.StructureKind.Module,
        name: complexType.typeName,
        isExported: true,
        statements: [propertyMetadata(complexType)]
    };
}
exports.complexTypeNamespace = complexTypeNamespace;
function getPropertyMetadataInitializer(complexType) {
    return `[${complexType.properties
        .map(property => `{
        originalName: '${property.originalName}',
        name: '${property.instancePropertyName}',
        type: ${property.isComplex ? property.jsType : `'${property.edmType}'`},
        isCollection: ${property.isCollection}
      }`)
        .join(', ')}]`;
}
function propertyMetadata(complexType) {
    return {
        kind: ts_morph_1.StructureKind.VariableStatement,
        declarationKind: ts_morph_1.VariableDeclarationKind.Const,
        declarations: [
            {
                name: '_propertyMetadata',
                initializer: getPropertyMetadataInitializer(complexType),
                type: `PropertyMetadata<${complexType.typeName}>[]`
            }
        ],
        docs: [
            `${util_1.unixEOL}Metadata information on all properties of the \`${complexType.typeName}\` complex type.`
        ],
        isExported: true
    };
}
