// @sap-cloud-sdk/generator/dist/service/package-json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.packageJson = void 0;
const util_1 = require("@sap-cloud-sdk/util");
const internal_1 = require("@sap-cloud-sdk/generator-common/internal");
/**
 * Generate the package.json for an odata client so it can be released as an npm module.
 * @param options - Options to generate the package.json
 * @returns The package.json contents.
 * @internal
 */
async function packageJson(options) {
    const oDataModule = options.oDataVersion === 'v2'
        ? '@sap-cloud-sdk/odata-v2'
        : '@sap-cloud-sdk/odata-v4';
    return (JSON.stringify({
        ...(0, internal_1.packageJsonBase)(options),
        files: [
            '**/*.js',
            '**/*.js.map',
            '**/*.d.ts',
            '**/d.ts.map',
            '**/*-csn.json'
        ],
        scripts: {
            compile: 'npx tsc'
        },
        dependencies: {
            '@sap-cloud-sdk/odata-common': `^${options.sdkVersion}`,
            [oDataModule]: `^${options.sdkVersion}`
        },
        devDependencies: {
            typescript: '~4.5'
        }
    }, null, 2) + util_1.unixEOL);
}
exports.packageJson = packageJson;
