// @sap-cloud-sdk/generator/dist/edmx-to-vdm/description-util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.entityDescription = exports.actionImportDescription = exports.functionImportDescription = exports.parameterDescription = exports.propertyDescription = exports.shortPropertyDescription = exports.longDescription = void 0;
const util_1 = require("@sap-cloud-sdk/util");
const generator_utils_1 = require("../generator-utils");
/**
 * @internal
 */
function longDescription(documented, described) {
    let docs = '';
    if (documented.Documentation) {
        const summary = (0, generator_utils_1.ensureString)(documented.Documentation.Summary);
        const longDesc = (0, generator_utils_1.ensureString)(documented.Documentation.LongDescription);
        docs = `${summary}${util_1.unixEOL}${longDesc}`.trim();
    }
    if (!docs && described) {
        docs = (0, generator_utils_1.ensureString)(described.description);
    }
    return (0, generator_utils_1.endWithDot)(docs.trim());
}
exports.longDescription = longDescription;
/**
 * @internal
 */
function shortPropertyDescription(property, swaggerProperty) {
    const desc = property['sap:quickinfo'] ||
        property['sap:label'] ||
        swaggerProperty?.title ||
        '';
    return (0, generator_utils_1.endWithDot)(desc.trim());
}
exports.shortPropertyDescription = shortPropertyDescription;
/**
 * @internal
 */
function propertyDescription(property, swaggerProperty) {
    const short = shortPropertyDescription(property, swaggerProperty);
    const long = longDescription(property, swaggerProperty);
    return `${short}${util_1.unixEOL}${long}`.trim();
}
exports.propertyDescription = propertyDescription;
/**
 * @internal
 */
function parameterDescription(parameter, swaggerParameter) {
    const short = (0, generator_utils_1.endWithDot)((0, util_1.titleFormat)(parameter.Name));
    const long = longDescription(parameter, swaggerParameter);
    return (0, generator_utils_1.endWithDot)((long || short).trim());
}
exports.parameterDescription = parameterDescription;
/**
 * @internal
 */
function functionImportDescription(swaggerDefinition, originalName) {
    return (0, generator_utils_1.endWithDot)(swaggerDefinition?.summary || (0, util_1.titleFormat)(originalName));
}
exports.functionImportDescription = functionImportDescription;
/**
 * @internal
 */
function actionImportDescription(swaggerDefinition, originalName) {
    return functionImportDescription(swaggerDefinition, originalName);
}
exports.actionImportDescription = actionImportDescription;
/**
 * @internal
 */
function entityDescription(entity, className) {
    return (entity.entityType['sap:label'] ||
        entity.swaggerDefinition?.title ||
        (0, util_1.titleFormat)(className));
}
exports.entityDescription = entityDescription;
