// @sap-cloud-sdk/generator/dist/edmx-to-vdm/common/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformOperationBase = void 0;
const util_1 = require("@sap-cloud-sdk/util");
const description_util_1 = require("../description-util");
const operation_parameter_1 = require("./operation-parameter");
/**
 * @internal
 * This transforms an EDMX operation (function or action) to its representation for the VDM.
 */
function transformOperationBase(edmxOperation, edmxParameters, type, swaggerDefinition, formatter, edmxBindingEntitySetName) {
    const operationName = edmxBindingEntitySetName
        ? formatter.originalToBoundOperationName(edmxBindingEntitySetName, edmxOperation.Name)
        : formatter.originalToOperationName(edmxOperation.Name);
    const operation = {
        originalName: edmxOperation.Name,
        name: operationName,
        parametersTypeName: (0, util_1.pascalCase)(`${operationName}Parameters`)
    };
    const parameters = (0, operation_parameter_1.getOperationParameters)(edmxOperation, edmxParameters, swaggerDefinition, formatter);
    return {
        ...operation,
        parameters,
        description: (0, description_util_1.functionImportDescription)(swaggerDefinition, operation.originalName),
        type,
        isBound: !!edmxBindingEntitySetName
    };
}
exports.transformOperationBase = transformOperationBase;
