// @sap-cloud-sdk/generator-common/dist/ts-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readCustomTsConfig = exports.formatTsConfig = exports.defaultTsConfig = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const util_1 = require("@sap-cloud-sdk/util");
const { readFile, lstat } = fs_1.promises;
/**
 * @internal
 */
exports.defaultTsConfig = {
    compilerOptions: {
        target: 'es2021',
        module: 'commonjs',
        lib: ['esnext'],
        declaration: true,
        declarationMap: false,
        sourceMap: true,
        diagnostics: true,
        moduleResolution: 'node',
        esModuleInterop: true,
        inlineSources: false,
        strict: true
    },
    include: ['**/*.ts'],
    exclude: ['dist/**/*', 'test/**/*', '**/*.spec.ts', 'node_modules/**/*']
};
/**
 * @internal
 */
function formatTsConfig() {
    return JSON.stringify(exports.defaultTsConfig, null, 2) + util_1.unixEOL;
}
exports.formatTsConfig = formatTsConfig;
/**
 * @internal
 */
async function readCustomTsConfig(configPath) {
    try {
        if ((await lstat(configPath)).isDirectory()) {
            configPath = (0, path_1.resolve)(configPath, 'tsconfig.json');
        }
        return await readFile(configPath, 'utf8');
    }
    catch (err) {
        throw new util_1.ErrorWithCause(`Could not read tsconfig.json at ${configPath}.`, err);
    }
}
exports.readCustomTsConfig = readCustomTsConfig;
