// @sap-cloud-sdk/generator/dist/edmx-parser/v2/edmx-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseFunctionImportsV2 = exports.parseAssociationSets = exports.parseAssociation = exports.parseEntityTypes = exports.parseEntitySetsV2 = exports.parseComplexTypesV2 = void 0;
const edmx_parser_1 = require("../common/edmx-parser");
const generator_utils_1 = require("../../generator-utils");
/**
 * @internal
 */
function parseComplexTypesV2(root) {
    return (0, edmx_parser_1.parseComplexTypesBase)(root);
}
exports.parseComplexTypesV2 = parseComplexTypesV2;
/**
 * @internal
 */
function parseEntitySetsV2(root) {
    return (0, edmx_parser_1.parseEntitySetsBase)(root);
}
exports.parseEntitySetsV2 = parseEntitySetsV2;
/**
 * @internal
 */
function parseEntityTypes(root) {
    return (0, edmx_parser_1.parseEntityTypesBase)(root);
}
exports.parseEntityTypes = parseEntityTypes;
/**
 * @internal
 */
function parseAssociation(root) {
    return (0, edmx_parser_1.getMergedPropertyWithNamespace)(root, 'Association');
}
exports.parseAssociation = parseAssociation;
/**
 * @internal
 */
function parseAssociationSets(root) {
    return (0, edmx_parser_1.getPropertyFromEntityContainer)(root, 'AssociationSet');
}
exports.parseAssociationSets = parseAssociationSets;
/**
 * @internal
 */
function parseFunctionImportsV2(root) {
    return (0, edmx_parser_1.getPropertyFromEntityContainer)(root, 'FunctionImport').map(f => ({
        ...f,
        Parameter: (0, generator_utils_1.forceArray)(f.Parameter)
    }));
}
exports.parseFunctionImportsV2 = parseFunctionImportsV2;
