// @sap-cloud-sdk/odata-common/dist/request/odata-create-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataCreateRequestConfig = void 0;
const odata_request_config_1 = require("./odata-request-config");
/**
 * OData create request configuration for an entity type.
 * @typeParam EntityT - Type of the entity to setup a request for
 */
class ODataCreateRequestConfig extends odata_request_config_1.ODataRequestConfig {
    /**
     * Creates an instance of ODataRequest.
     * @param entityApi - Entity API for building and executing the request.
     * @param oDataUri - URI conversion functions.
     */
    constructor(entityApi, oDataUri) {
        super('post', entityApi.entityConstructor._defaultBasePath);
        this.entityApi = entityApi;
        this.oDataUri = oDataUri;
    }
    resourcePath() {
        return this.parentKeys === undefined
            ? this.entityApi.entityConstructor._entityName
            : this.resourcePathAsChild();
    }
    queryParameters() {
        return this.prependDollarToQueryParameters({});
    }
    resourcePathAsChild() {
        return (this.oDataUri.getResourcePathForKeys(this.parentKeys, this.childField._entityApi) +
            '/' +
            this.childField._fieldName);
    }
}
exports.ODataCreateRequestConfig = ODataCreateRequestConfig;
