// @sap-cloud-sdk/odata-v4/dist/request/odata-bound-function-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OdataBoundFunctionImportRequestConfig = exports.ODataBoundFunctionRequestConfig = void 0;
const odata_function_request_config_1 = require("./odata-function-request-config");
/**
 * Function request configuration for an entity type.
 * @typeParam DeSerializersT - Type of the deserializer use on the request
 * @typeParam ParametersT - Type of the parameter to setup a request with
 */
class ODataBoundFunctionRequestConfig extends odata_function_request_config_1.ODataFunctionRequestConfig {
    constructor(method, entityApi, functionName, parameters, oDataUri) {
        super(method, entityApi.entityConstructor._defaultBasePath, functionName, parameters, oDataUri);
        this.entityApi = entityApi;
        this.oDataUri = oDataUri;
    }
    resourcePath() {
        return `${this.oDataUri.getResourcePathForKeys(this.keys, this.entityApi)}/${super.resourcePath()}`;
    }
    queryParameters() {
        return {};
    }
}
exports.ODataBoundFunctionRequestConfig = ODataBoundFunctionRequestConfig;
/**
 * @deprecated Since 3.3.0. Use {@link ODataBoundFunctionRequestConfig} instead.
 */
exports.OdataBoundFunctionImportRequestConfig = ODataBoundFunctionRequestConfig;
