// @sap-cloud-sdk/odata-v4/dist/request/odata-bound-action-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataBoundActionImportRequestConfig = exports.ODataBoundActionRequestConfig = void 0;
const odata_action_request_config_1 = require("./odata-action-request-config");
/**
 * Action request configuration for an entity type.
 * @typeParam DeSerializersT - Type of the deserializer use on the request
 * @typeParam ParametersT - Type of the parameter to setup a request with
 */
class ODataBoundActionRequestConfig extends odata_action_request_config_1.ODataActionRequestConfig {
    constructor(entityApi, actionName, parameters, oDataUri) {
        super(entityApi.entityConstructor._defaultBasePath, actionName, parameters, oDataUri);
        this.entityApi = entityApi;
        this.oDataUri = oDataUri;
    }
    resourcePath() {
        return `${this.oDataUri.getResourcePathForKeys(this.keys, this.entityApi)}/${super.resourcePath()}`;
    }
    queryParameters() {
        return {};
    }
}
exports.ODataBoundActionRequestConfig = ODataBoundActionRequestConfig;
/**
 * @deprecated Since 3.3.0. Use {@link ODataBoundActionRequestConfig} instead.
 */
class ODataBoundActionImportRequestConfig extends ODataBoundActionRequestConfig {
    constructor(
    /**
     * @deprecated Since 3.3.0. This parameter is unused and will be removed in the next major release.
     */
    method, entityApi, actionName, parameters, oDataUri) {
        super(entityApi, actionName, parameters, oDataUri);
        this.entityApi = entityApi;
        this.oDataUri = oDataUri;
    }
}
exports.ODataBoundActionImportRequestConfig = ODataBoundActionImportRequestConfig;
