// @sap-cloud-sdk/odata-v4/dist/request/odata-action-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataActionImportRequestConfig = exports.ODataActionRequestConfig = void 0;
const odata_common_1 = require("@sap-cloud-sdk/odata-common");
/**
 * Action request configuration for an entity type.
 * @typeParam DeSerializersT - Type of the deserializer use on the request
 * @typeParam ParametersT - Type of the parameter to setup a request with
 */
class ODataActionRequestConfig extends odata_common_1.ODataRequestConfig {
    /**
     * Creates an instance of ODataActionRequestConfig.
     * @param defaultBasePath - Default base path of the service.
     * @param actionName - The name of the action.
     * @param parameters - Parameters of the action.
     * @param oDataUri - URI conversion functions.
     */
    constructor(defaultBasePath, actionName, parameters, oDataUri) {
        super('post', defaultBasePath);
        this.actionName = actionName;
        this.parameters = parameters;
        this.oDataUri = oDataUri;
        this.payload = this.buildHttpPayload(parameters);
    }
    resourcePath() {
        return this.actionName;
    }
    queryParameters() {
        return {};
    }
    buildHttpPayload(parameters) {
        const payload = Object.keys(parameters).reduce((all, key) => {
            const payloadElement = parameters[key];
            if (typeof payloadElement.value !== 'undefined') {
                all[payloadElement.originalName] = payloadElement.value;
            }
            return all;
        }, {});
        return payload;
    }
}
exports.ODataActionRequestConfig = ODataActionRequestConfig;
/**
 * @deprecated Since 3.3.0. Use {@link ODataActionRequestConfig} instead.
 */
exports.ODataActionImportRequestConfig = ODataActionRequestConfig;
