// @sap-cloud-sdk/odata-common/dist/request/odata-delete-request-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataDeleteRequestConfig = void 0;
const odata_request_config_1 = require("./odata-request-config");
/**
 * OData delete request configuration for an entity type.
 * @typeParam EntityT - Type of the entity to setup a request for
 */
class ODataDeleteRequestConfig extends odata_request_config_1.ODataRequestConfig {
    /**
     * Creates an instance of ODataDeleteRequestConfig.
     * @param entityApi - Entity API for building and executing the request.
     * @param oDataUri - URI conversion functions.
     */
    constructor(entityApi, oDataUri) {
        super('delete', entityApi.entityConstructor._defaultBasePath);
        this.entityApi = entityApi;
        this.oDataUri = oDataUri;
        this.versionIdentifierIgnored = false;
    }
    resourcePath() {
        return this.oDataUri.getResourcePathForKeys(this.keys, this.entityApi);
    }
    queryParameters() {
        return this.prependDollarToQueryParameters({});
    }
}
exports.ODataDeleteRequestConfig = ODataDeleteRequestConfig;
